`marker orient="-1"` does not orient correctly (uses 'auto')

Ahmad-S792 opened the pull request:
#### 6bf2a1e9acdca0b3b54a5cb6332ecbf591ff565f
<pre>
`marker orient=&quot;-1&quot;` does not orient correctly (uses &apos;auto&apos;)

<a href="https://bugs.webkit.org/show_bug.cgi?id=256670">https://bugs.webkit.org/show_bug.cgi?id=256670</a>

Reviewed by Said Abou-Hallawa.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Merge: <a href="https://chromium.googlesource.com/chromium/src.git/+/886546446df947816927a7757de45fab1646c919">https://chromium.googlesource.com/chromium/src.git/+/886546446df947816927a7757de45fab1646c919</a>

The value -1 (degrees) is a valid angle, so using it to indicate that
&apos;auto&apos; orientation should be used does not work out.
Just check &apos;orientType&apos; directly instead and simplify the angle getter.

* Source/WebCore/rendering/svg/RenderSVGResourceMarker.cpp:
(RenderSVGResourceMarker::angle): Simplify logic
(RenderSVGResourceMarker::markerTransformation): Call &apos;orient&apos; directly
* Source/WebCore/rendering/svg/RenderSVGResourceMarker.h: Update &apos;float angle&apos; to std::optional
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp: Add Text Tree dump support
* LayoutTests/svg/markers/marker-orientation-minus-one.html: Add Test Case
* LayoutTests/svg/markers/marker-orientation-minus-one-expected.html: Add Test Case Expectation

Canonical link: <a href="https://commits.webkit.org/264053@main">https://commits.webkit.org/264053@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/12c4debdc72640986a112560b267957d88115ece

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6544 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6756 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6936 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8125 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6828 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6541 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/7216 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6705 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9724 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6659 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6801 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5943 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8213 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/4264 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5912 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13750 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/6231 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5992 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/8391 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6477 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5298 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5878 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/5840 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1543 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10042 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6250 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/rendering/svg/RenderSVGResourceMarker.cpp:
```diff
@@ -94,21 +96,16 @@ FloatPoint RenderSVGResourceMarker::referencePoint() const
-
-    return angle;
+    return markerElement().orientAngle().value();
 }
```
I would suggest making this function return `std::optional<float>` since the `markerElement()` is not necessarily has an angle.
```
std::optional<float> RenderSVGResourceMarker::angle() const
{
    if (markerElement().orientType() == SVGMarkerOrientAngle)
        return markerElement().orientAngle().value();
    return std::nullopt;
}
```

Ahmad-S792 replied:
Done and added stuff in `.h` accordingly as well.

shallawa commented on Source/WebCore/rendering/svg/RenderSVGResourceMarker.cpp:
```diff
@@ -94,21 +96,16 @@ FloatPoint RenderSVGResourceMarker::referencePoint() const
     AffineTransform transform;
     transform.translate(origin);
-    transform.rotate(markerAngle == -1 ? autoAngle : markerAngle);
+    transform.rotate(markerElement().orientType() == SVGMarkerOrientAngle ? angle() : autoAngle);
```
If `angle()` returns `std::optional<float>` this statement can be simplified like this:
```
    transform.rotate(angle().value_or(autoAngle));
```

Ahmad-S792 replied:
Done.

shallawa commented on Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
```diff
@@ -488,7 +489,7 @@ void writeSVGResourceContainer(TextStream& ts, const RenderSVGResourceContainer&
+        if (marker.markerElement().orientType() != SVGMarkerOrientAngle)
             ts << "auto" << "]\n";
         else
             ts << marker.angle() << "]\n";
```
This also can be simplified like this:
```
    if (auto angle = this->angle())
        ts << *angle << "]\n";
    else
        ts << "auto" << "]\n";
```

Ahmad-S792 replied:
@shallawa - this change is leading to build failure. I fixed others and also tried to fix this twice as per my knowledge: This is build error:
```SVGRenderTreeAsText.cpp:492:26: error: invalid use of 'this' outside of a non-static member function```

Ahmad-S792 replied:
I have reverted to original, I tried to fix it but it seems to be not working.

shallawa replied:
Sorry it is my mistake. I did not realize this change is in a different source file. I think this should work:
```
    if (auto angle = marker.angle())
        ts << *angle << "]\n";
    else
        ts << "auto" << "]\n";
```

Ahmad-S792 replied:
Pushed new build with suggested change. :-) Thanks for your help as always. @shallawa - one less bug in WebKit now.

Ahmad-S792 committed:
`marker orient="-1"` does not orient correctly (uses 'auto')

https://bugs.webkit.org/show_bug.cgi?id=256670

Reviewed by Said Abou-Hallawa.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Merge: https://chromium.googlesource.com/chromium/src.git/+/886546446df947816927a7757de45fab1646c919

The value -1 (degrees) is a valid angle, so using it to indicate that
'auto' orientation should be used does not work out.
Just check 'orientType' directly instead and simplify the angle getter.

* Source/WebCore/rendering/svg/RenderSVGResourceMarker.cpp:
(RenderSVGResourceMarker::angle): Simplify logic
(RenderSVGResourceMarker::markerTransformation): Call 'orient' directly
* Source/WebCore/rendering/svg/RenderSVGResourceMarker.h: Update 'float angle' to std::optional
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp: Add Text Tree dump support
* LayoutTests/svg/markers/marker-orientation-minus-one.html: Add Test Case
* LayoutTests/svg/markers/marker-orientation-minus-one-expected.html: Add Test Case Expectation

Canonical link: https://commits.webkit.org/264053@main
